### ISSUE TITLE
With Restricted Image Loads, Invalid Spirv Generated

### ISSUE THREAD

**Issue by cwfitzgerald** [gfx-rs/wgpu#4581]
Run `gpu_tests::clear_texture::clear_texture_depth` with image_load bounds checks forced to restricted.

```
[2023-10-27T06:12:28Z ERROR wgpu_hal::vulkan::instance] VALIDATION [VUID-VkShaderModuleCreateInfo-pCode-01379 (0x2a1bf17f)]
        Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-01379 ] Object 0: handle = 0x21b47616880, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x2a1bf17f | SPIR-V module not valid: OpConstantComposite Constituent <id> '11[%11]'s type does not match Result Type <id> '33[%33]'s vector element type.
      %89 = OpConstantComposite %v3uint %int_1 %int_1 %int_1
     The Vulkan spec states: If pCode is a pointer to GLSL code, it must be valid GLSL code written to the GL_KHR_vulkan_glsl GLSL extension specification (https://vulkan.lunarg.com/doc/view/1.3.250.0/windows/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-01379)
```

Will fill in a little more information soon.

**Comment by cwfitzgerald** [comment]
I believe this (or very similar issue) is blocking my work's upgrade to 0.18. Going to investigate a bit more.

**Comment by jimblandy** [comment]
Given that this is a type error building a constant, it's likely to be fallout from the WGSL const expressions work. It probably won't have much to actually do with image bounds checks.

If you can find a reduced test case, this should be not too involved to fix.

**Comment by cwfitzgerald** [comment]
Alright, this can reproduce in minimal code through the following wgsl:

```rust
fn f() {
    let x = vec4f(vec3f(1.0), 1.0);
}
```

Which generates:

```
    %float_1 = OpConstant %float 1
         %11 = OpConstantComposite %v3float %float_1 %float_1 %float_1
         %12 = OpConstantComposite %v4float %11 %float_1
```

But `OpConstantComposite`, unlike `OpCompositeConstruct` can't be used with arbitrary combinations of values as long as you have enough of them. It can only be used with floats.

**Comment by teoxoy** [comment]
Hmm, I remember we ran into this in the const-expr PR but thought we fixed it.

**Comment by jimblandy** [comment]
I think it's because we didn't handle `Splat`.

**Comment by LPGhatguy** [comment]
I hit this bug recently! @cwfitzgerald asked me to post some additional context in case other folks hit this in the future from the same path I did.

On Nvidia driver 546.33, the invalid SPIR-V generated by naga 0.14.0 caused the driver to print to my application's output and triggers a device lost:

> NVVM compilation failed: 1
> [ERROR wgpu::backend::direct] Handling wgpu errors as fatal by default
> [ERROR crash_handler] Fatal error: The game crashed! Details will be written to crash.json.
>
>    Details: wgpu error: Validation Error
>
>    Caused by:
>        In Device::create_render_pipeline
>          note: label = `Render Sky Cube`
>        Parent device is lost
>
>
>    Location: file C:\Users\lucien\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wgpu-0.18.0\src\backend\direct.rs on line 3111

Searching for `NVVM compilation failed: 1`, I found exactly one result, [this post on the Nvidia forums](https://forums.developer.nvidia.com/t/vkcreategraphicspipelines-call-fail-to-compile-specific-spir-v-fragment-shader-on-some-nvidia-laptop-gpus/273687).

Installing the beta driver they mention, 538.03, changes the error message. With that driver version, the driver exits my program (with exit code 0) after printing:

> LLVM ERROR: Type mismatch in constant table!

This must be an internal driver compilation failure, but it's funny that it surfaces by appearing like a successful exit of the program.

In any case, upgrading to naga 0.14.2 fixes the issue regardless of driver version.
